src/main/java/com/google/changestreams/sample/bigquery/changelog/fullschema/SpannerChangeStreamsToBigQuery.java

```diff
@@ -0,0 +1,329 @@
+      .apply(
+        "Write Failed Mod JSON To DLQ",
+        DLQWriteTransform.WriteDLQ.newBuilder()
+          .withDlqDirectory(dlqDirectory)
+          .withTmpDirectory(tempDlqDirectory)
+          .build());
+
+    PCollection<FailsafeElement<String, String>> nonRetryableDlqModJsonFailsafe =
+      dlqModJson
+        .get(DeadLetterQueueManager.PERMANENT_ERRORS)
+        .setCoder(FAILSAFE_ELEMENT_CODER);
+
+    nonRetryableDlqModJsonFailsafe
+      .apply(
+        "Write Mod JSON With Non-retryable Error To DLQ",
+        MapElements.via(new StringDeadLetterQueueSanitizer()))
+      .setCoder(StringUtf8Coder.of())
+      .apply(
+        DLQWriteTransform.WriteDLQ.newBuilder()
+          .withDlqDirectory(dlqManager.getSevereDlqDirectoryWithDateTime())
+          .withTmpDirectory(dlqManager.getSevereDlqDirectory() + "tmp/")
+          .build());
+
+    return pipeline.run();
+  }
+
+  private static DeadLetterQueueManager buildDlqManager(Options options) {
+    final String tempLocation =
+      options.as(DataflowPipelineOptions.class).getTempLocation().endsWith("/")
+        ? options.as(DataflowPipelineOptions.class).getTempLocation()
+        : options.as(DataflowPipelineOptions.class).getTempLocation() + "/";
+
+    final String dlqDirectory =
+      options.getDeadLetterQueueDirectory().isEmpty()
+        ? tempLocation + "dlq/"
+        : options.getDeadLetterQueueDirectory();
+
+    LOG.info("Dead letter queue directory: {}", dlqDirectory);
+    return DeadLetterQueueManager.create(dlqDirectory);
+  }
+
+  private static String getBigQueryProject(Options options) {
+    return options.getBigQueryProjectId().isEmpty()
+      ? options.getProject()
+      : options.getBigQueryProjectId();
+  }
+
+  /**
+   * Remove the following intermediate metadata fields that are not user data from TableRow:
+   * * _metadata_error.
+   * * _metadata_retry_count.
+   * * _metadata_spanner_original_payload_json.
+   */
+  private static TableRow removeIntermediateMetadataFields(TableRow tableRow) {
+    TableRow cleanTableRow = tableRow.clone();
+    final Set<String> rowKeys = tableRow.keySet();
+    final Set<String> metadataFields = SchemaUtils.getBigQueryIntermediateMetadataFieldNames();
+
+    for (String rowKey : rowKeys) {
+      if (metadataFields.contains(rowKey)) {
+        cleanTableRow.remove(rowKey);
+      }
+    }
+
+    return cleanTableRow;
+  }
+
+  // ModWithMetadata Json string is the original message that can be consumed by the pipeline.
+  static class DataChangeRecordToModJsonFn extends DoFn<DataChangeRecord, String> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DataChangeRecordToModJsonFn.class);
+
+    @ProcessElement
+    public void process(@Element DataChangeRecord input,
+                        OutputReceiver<String> receiver) {
+      for (org.apache.beam.sdk.io.gcp.spanner.changestreams.model.Mod changeStreamsMod : input.getMods()) {
```

Comment:
This might be a cleaner way to write this function: https://github.com/nancyxu123/change-stream-sample/blob/main/src/main/java/com/google/changestreams/sample/timestampOrdering/DataflowOrderedGloballyByKeyCodeSample.java#L207

Reply:
Thanks, that's a cleaner way if we do something simple in the for loop, with exception handling the code becomes less readable IMO, I can make the change if you think that's fine though.